### ISSUE TITLE
TSAN lock cycle detected in application UT

### ISSUE THREAD

**Issue by jwalch** [openssl/openssl#13634]
TSAN is detecting a couple of seemingly related lock order inversion cycles in my application's unit test. I believe it has been present since alpha7. Unfortunately, it is proprietary code and I'm not at liberty to share the source at this time. Hopefully, however, the TSAN backtrace outputs (minus those frames) are sufficient to illustrate the issue and facilitate reproduction.

EDIT: Please note that source line numbers may not align perfectly due to ocassional OpenSSL source modifications on 
our end (which I believe are not relevant to this issue).

```
WARNING: ThreadSanitizer: lock-order-inversion (potential deadlock) (pid=71331)
  Cycle in lock order graph: M768 (0x7b1000002280) => M811 (0x7b1000002ac0) => M1014 (0x7b10000045c0) => M768

  Mutex M811 acquired here while holding mutex M768 in main thread:
    [...]
    #1 CRYPTO_THREAD_write_lock crypto/threads_pthread.c:89 (libcrypto.so.3+0x4a7668)
    #2 ossl_provider_find crypto/provider_core.c:230 (libcrypto.so.3+0x49f351)
    #3 provider_conf_load crypto/provider_conf.c:118 (libcrypto.so.3+0x49e9e5)
    #4 provider_conf_init crypto/provider_conf.c:168 (libcrypto.so.3+0x49e5f6)
    #5 module_init crypto/conf/conf_mod.c:438 (libcrypto.so.3+0x2e0cf7)
    #6 module_run crypto/conf/conf_mod.c:294 (libcrypto.so.3+0x2deb82)
    #7 CONF_modules_load crypto/conf/conf_mod.c:193 (libcrypto.so.3+0x2de8ae)
    #8 CONF_modules_load_file_ex crypto/conf/conf_mod.c:236 (libcrypto.so.3+0x2ded4b)
    #9 CONF_modules_load_file crypto/conf/conf_mod.c:258 (libcrypto.so.3+0x2df02b)
    #10 openssl_config_int crypto/conf/conf_sap.c:63 (libcrypto.so.3+0x2e13b9)
    #11 ossl_init_config crypto/init.c:237 (libcrypto.so.3+0x486895)
    #12 ossl_init_config_ossl_ crypto/init.c:235 (libcrypto.so.3+0x485ff2)
    [...]
    #14 CRYPTO_THREAD_run_once crypto/threads_pthread.c:129 (libcrypto.so.3+0x4a7825)
    #15 OPENSSL_init_crypto crypto/init.c:544 (libcrypto.so.3+0x4857a5)
    #16 OPENSSL_init_ssl ssl/ssl_init.c:131 (libssl.so.3+0x7b184)
    [...]

    Hint: use TSAN_OPTIONS=second_deadlock_stack=1 to get more informative warning message

  Mutex M1014 acquired here while holding mutex M811 in main thread:
    [...]
    #1 CRYPTO_THREAD_write_lock crypto/threads_pthread.c:89 (libcrypto.so.3+0x4a7668)
    #2 ossl_namemap_add_names crypto/core_namemap.c:280 (libcrypto.so.3+0x47f73d)
    #3 construct_evp_method crypto/evp/evp_fetch.c:195 (libcrypto.so.3+0x41c20b)
    #4 ossl_method_construct_this crypto/core_fetch.c:70 (libcrypto.so.3+0x47e36a)
    #5 algorithm_do_this crypto/core_algorithm.c:65 (libcrypto.so.3+0x47dd9b)
    #6 provider_forall_loaded crypto/provider_core.c:675 (libcrypto.so.3+0x4a2051)
    #7 ossl_provider_forall_loaded crypto/provider_core.c:753 (libcrypto.so.3+0x4a1cdb)
    #8 ossl_algorithm_do_all crypto/core_algorithm.c:109 (libcrypto.so.3+0x47da8a)
    #9 ossl_method_construct crypto/core_fetch.c:124 (libcrypto.so.3+0x47e06a)
    #10 inner_evp_generic_fetch crypto/evp/evp_fetch.c:320 (libcrypto.so.3+0x41aa0b)
    #11 evp_generic_fetch crypto/evp/evp_fetch.c:367 (libcrypto.so.3+0x41a613)
    #12 EVP_CIPHER_fetch crypto/evp/evp_enc.c:1499 (libcrypto.so.3+0x412893)
    #13 ssl_evp_cipher_fetch ssl/ssl_lib.c:6526 (libssl.so.3+0xa1b5f)
    #14 ssl_load_ciphers ssl/ssl_ciph.c:345 (libssl.so.3+0x6b0e0)
    #15 SSL_CTX_new_ex ssl/ssl_lib.c:3332 (libssl.so.3+0x91022)
    #16 SSL_CTX_new ssl/ssl_lib.c:3508 (libssl.so.3+0x921cc)
    [...]

  Mutex M768 acquired here while holding mutex M1014 in main thread:
    [...]
    #1 CRYPTO_THREAD_write_lock crypto/threads_pthread.c:89 (libcrypto.so.3+0x4a7668)
    #2 OPENSSL_init_crypto crypto/init.c:542 (libcrypto.so.3+0x48576d)
    #3 OBJ_sn2nid crypto/objects/obj_dat.c:615 (libcrypto.so.3+0x4cf5a1)
    #4 evp_pkey_name2type crypto/evp/p_lib.c:983 (libcrypto.so.3+0x44a0aa)
    #5 help_get_legacy_alg_type_from_keymgmt crypto/evp/pmeth_lib.c:162 (libcrypto.so.3+0x45f0d9)
    #6 do_name crypto/core_namemap.c:126 (libcrypto.so.3+0x47ec52)
    #7 doall_util_fn crypto/lhash/lhash.c:207 (libcrypto.so.3+0x47a0ba)
    #8 OPENSSL_LH_doall_arg crypto/lhash/lhash.c:222 (libcrypto.so.3+0x47a1a0)
    #9 lh_NAMENUM_ENTRY_doall_DOALL_NAMES_DATA crypto/core_namemap.c:129 (libcrypto.so.3+0x47eb35)
    #10 ossl_namemap_doall_names crypto/core_namemap.c:141 (libcrypto.so.3+0x47ea7e)
    #11 evp_names_do_all crypto/evp/evp_fetch.c:574 (libcrypto.so.3+0x41ba4b)
    #12 EVP_KEYMGMT_names_do_all crypto/evp/keymgmt_meth.c:277 (libcrypto.so.3+0x435c82)
    #13 get_legacy_alg_type_from_keymgmt crypto/evp/pmeth_lib.c:169 (libcrypto.so.3+0x45f037)
    #14 int_ctx_new crypto/evp/pmeth_lib.c:277 (libcrypto.so.3+0x45310d)
    #15 EVP_PKEY_CTX_new_from_pkey crypto/evp/pmeth_lib.c:351 (libcrypto.so.3+0x453607)
    #16 ssl_setup_sig_algs ssl/t1_lib.c:1175 (libssl.so.3+0xbaf96)
    #17 SSL_CTX_new_ex ssl/ssl_lib.c:3335 (libssl.so.3+0x91045)
    #18 SSL_CTX_new ssl/ssl_lib.c:3508 (libssl.so.3+0x921cc)
    [...]

SUMMARY: ThreadSanitizer: lock-order-inversion (potential deadlock) crypto/threads_pthread.c:89 in CRYPTO_THREAD_write_lock
==================
==================
WARNING: ThreadSanitizer: lock-order-inversion (potential deadlock) (pid=71331)
  Cycle in lock order graph: M768 (0x7b1000002280) => M811 (0x7b1000002ac0) => M768

  Mutex M811 acquired here while holding mutex M768 in main thread:
    [...]
    #1 CRYPTO_THREAD_write_lock crypto/threads_pthread.c:89 (libcrypto.so.3+0x4a7668)
    #2 ossl_provider_find crypto/provider_core.c:230 (libcrypto.so.3+0x49f351)
    #3 provider_conf_load crypto/provider_conf.c:118 (libcrypto.so.3+0x49e9e5)
    #4 provider_conf_init crypto/provider_conf.c:168 (libcrypto.so.3+0x49e5f6)
    #5 module_init crypto/conf/conf_mod.c:438 (libcrypto.so.3+0x2e0cf7)
    #6 module_run crypto/conf/conf_mod.c:294 (libcrypto.so.3+0x2deb82)
    #7 CONF_modules_load crypto/conf/conf_mod.c:193 (libcrypto.so.3+0x2de8ae)
    #8 CONF_modules_load_file_ex crypto/conf/conf_mod.c:236 (libcrypto.so.3+0x2ded4b)
    #9 CONF_modules_load_file crypto/conf/conf_mod.c:258 (libcrypto.so.3+0x2df02b)
    #10 openssl_config_int crypto/conf/conf_sap.c:63 (libcrypto.so.3+0x2e13b9)
    #11 ossl_init_config crypto/init.c:237 (libcrypto.so.3+0x486895)
    #12 ossl_init_config_ossl_ crypto/init.c:235 (libcrypto.so.3+0x485ff2)
    [...]
    #14 CRYPTO_THREAD_run_once crypto/threads_pthread.c:129 (libcrypto.so.3+0x4a7825)
    #15 OPENSSL_init_crypto crypto/init.c:544 (libcrypto.so.3+0x4857a5)
    #16 OPENSSL_init_ssl ssl/ssl_init.c:131 (libssl.so.3+0x7b184)
    [...]

    Hint: use TSAN_OPTIONS=second_deadlock_stack=1 to get more informative warning message

  Mutex M768 acquired here while holding mutex M811 in main thread:
    [...]
    #1 CRYPTO_THREAD_write_lock crypto/threads_pthread.c:89 (libcrypto.so.3+0x4a7668)
    #2 OPENSSL_init_crypto crypto/init.c:542 (libcrypto.so.3+0x48576d)
    #3 ossl_method_store_fetch crypto/property/property.c:341 (libcrypto.so.3+0x50ef96)
    #4 get_evp_method_from_store crypto/evp/evp_fetch.c:134 (libcrypto.so.3+0x41be76)
    #5 ossl_method_construct crypto/core_fetch.c:109 (libcrypto.so.3+0x47dfc1)
    #6 inner_evp_generic_fetch crypto/evp/evp_fetch.c:320 (libcrypto.so.3+0x41aa0b)
    #7 evp_generic_fetch crypto/evp/evp_fetch.c:367 (libcrypto.so.3+0x41a613)
    #8 EVP_KEYMGMT_fetch crypto/evp/keymgmt_meth.c:213 (libcrypto.so.3+0x4356f5)
    #9 add_provider_groups ssl/t1_lib.c:360 (libssl.so.3+0xc939b)
    #10 tls_group_capability providers/common/capabilities.c:175 (libcrypto.so.3+0x7087fa)
    #11 provider_get_capabilities providers/common/capabilities.c:185 (libcrypto.so.3+0x708746)
    #12 ossl_provider_get_capabilities crypto/provider_core.c:864 (libcrypto.so.3+0x4a2ac7)
    #13 OSSL_PROVIDER_get_capabilities crypto/provider.c:91 (libcrypto.so.3+0x49e1b5)
    #14 discover_provider_groups ssl/t1_lib.c:397 (libssl.so.3+0xb8b73)
    #15 provider_forall_loaded crypto/provider_core.c:675 (libcrypto.so.3+0x4a2051)
    #16 ossl_provider_forall_loaded crypto/provider_core.c:753 (libcrypto.so.3+0x4a1cdb)
    #17 OSSL_PROVIDER_do_all crypto/provider.c:127 (libcrypto.so.3+0x49e3c5)
    #18 ssl_load_groups ssl/t1_lib.c:403 (libssl.so.3+0xb8ae3)
    #19 SSL_CTX_new_ex ssl/ssl_lib.c:3339 (libssl.so.3+0x91068)
    #20 SSL_CTX_new ssl/ssl_lib.c:3508 (libssl.so.3+0x921cc)
    [...]

SUMMARY: ThreadSanitizer: lock-order-inversion (potential deadlock) crypto/threads_pthread.c:89 in CRYPTO_THREAD_write_lock
==================
OK! (2.186s)
ThreadSanitizer: reported 2 warnings
```

**Comment by jwalch** [comment]
I see a couple of other cycles that I believe are fundamentally the same thing as well, but including them for completeness:

```
WARNING: ThreadSanitizer: lock-order-inversion (potential deadlock) (pid=6699)
  Cycle in lock order graph: M699 (0x7b1000001fc0) => M742 (0x7b10000028c0) => M962 (0x7b10000043c0) => M699

  Mutex M742 acquired here while holding mutex M699 in main thread:
    [...]
    #1 CRYPTO_THREAD_write_lock crypto/threads_pthread.c:89 (libcrypto.so.3+0x4a7668)
    #2 ossl_provider_find crypto/provider_core.c:230 (libcrypto.so.3+0x49f351)
    #3 provider_conf_load crypto/provider_conf.c:118 (libcrypto.so.3+0x49e9e5)
    #4 provider_conf_init crypto/provider_conf.c:168 (libcrypto.so.3+0x49e5f6)
    #5 module_init crypto/conf/conf_mod.c:438 (libcrypto.so.3+0x2e0cf7)
    #6 module_run crypto/conf/conf_mod.c:294 (libcrypto.so.3+0x2deb82)
    #7 CONF_modules_load crypto/conf/conf_mod.c:193 (libcrypto.so.3+0x2de8ae)
    #8 CONF_modules_load_file_ex crypto/conf/conf_mod.c:236 (libcrypto.so.3+0x2ded4b)
    #9 CONF_modules_load_file crypto/conf/conf_mod.c:258 (libcrypto.so.3+0x2df02b)
    #10 openssl_config_int crypto/conf/conf_sap.c:63 (libcrypto.so.3+0x2e13b9)
    #11 ossl_init_config crypto/init.c:237 (libcrypto.so.3+0x486895)
    #12 ossl_init_config_ossl_ crypto/init.c:235 (libcrypto.so.3+0x485ff2)
    [...]
    #14 CRYPTO_THREAD_run_once crypto/threads_pthread.c:129 (libcrypto.so.3+0x4a7825)
    #15 OPENSSL_init_crypto crypto/init.c:544 (libcrypto.so.3+0x4857a5)
    #16 OBJ_obj2nid crypto/objects/obj_dat.c:364 (libcrypto.so.3+0x4cf0b5)
    #17 c2i_ASN1_OBJECT crypto/asn1/a_object.c:275 (libcrypto.so.3+0x1ad05e)
    #18 asn1_ex_c2i crypto/asn1/tasn_dec.c:805 (libcrypto.so.3+0x1d9a6d)
    #19 asn1_d2i_ex_primitive crypto/asn1/tasn_dec.c:762 (libcrypto.so.3+0x1d7721)
    #20 asn1_item_embed_d2i crypto/asn1/tasn_dec.c:187 (libcrypto.so.3+0x1d4ea9)
    #21 asn1_template_noexp_d2i crypto/asn1/tasn_dec.c:626 (libcrypto.so.3+0x1d8d7b)
    #22 asn1_template_ex_d2i crypto/asn1/tasn_dec.c:503 (libcrypto.so.3+0x1d6d36)
    #23 asn1_item_embed_d2i crypto/asn1/tasn_dec.c:369 (libcrypto.so.3+0x1d628a)
    #24 ASN1_item_ex_d2i crypto/asn1/tasn_dec.c:129 (libcrypto.so.3+0x1d4a40)
    #25 ASN1_item_d2i crypto/asn1/tasn_dec.c:119 (libcrypto.so.3+0x1d4909)
    #26 ASN1_item_dup crypto/asn1/a_dup.c:70 (libcrypto.so.3+0x1a7162)
    #27 X509_EXTENSION_dup crypto/x509/x_exten.c:28 (libcrypto.so.3+0x6614ed)
    [...]

    Hint: use TSAN_OPTIONS=second_deadlock_stack=1 to get more informative warning message

  Mutex M962 acquired here while holding mutex M742 in main thread:
    [...]
    #1 CRYPTO_THREAD_write_lock crypto/threads_pthread.c:89 (libcrypto.so.3+0x4a7668)
    #2 ossl_namemap_add_name_n crypto/core_namemap.c:254 (libcrypto.so.3+0x47f2dc)
    #3 ossl_namemap_add_name crypto/core_namemap.c:265 (libcrypto.so.3+0x47f63d)
    #4 construct_loader crypto/store/store_meth.c:241 (libcrypto.so.3+0x5bcbfb)
    #5 ossl_method_construct_this crypto/core_fetch.c:70 (libcrypto.so.3+0x47e36a)
    #6 algorithm_do_this crypto/core_algorithm.c:65 (libcrypto.so.3+0x47dd9b)
    #7 provider_forall_loaded crypto/provider_core.c:675 (libcrypto.so.3+0x4a2051)
    #8 ossl_provider_forall_loaded crypto/provider_core.c:753 (libcrypto.so.3+0x4a1cdb)
    #9 ossl_algorithm_do_all crypto/core_algorithm.c:109 (libcrypto.so.3+0x47da8a)
    #10 ossl_method_construct crypto/core_fetch.c:124 (libcrypto.so.3+0x47e06a)
    #11 inner_loader_fetch crypto/store/store_meth.c:295 (libcrypto.so.3+0x5bbe52)
    #12 OSSL_STORE_LOADER_fetch crypto/store/store_meth.c:317 (libcrypto.so.3+0x5bbc47)
    #13 OSSL_STORE_attach crypto/store/store_lib.c:940 (libcrypto.so.3+0x5bb3ac)
    #14 pem_read_bio_key crypto/pem/pem_pkey.c:57 (libcrypto.so.3+0x4eb82d)
    #15 PEM_read_bio_PrivateKey_ex crypto/pem/pem_pkey.c:144 (libcrypto.so.3+0x4ebd6a)
    #16 PEM_read_bio_PrivateKey crypto/pem/pem_pkey.c:151 (libcrypto.so.3+0x4ebe0e)
    [...]

  Mutex M699 acquired here while holding mutex M962 in main thread:
    [...]
    #1 CRYPTO_THREAD_write_lock crypto/threads_pthread.c:89 (libcrypto.so.3+0x4a7668)
    #2 OPENSSL_init_crypto crypto/init.c:542 (libcrypto.so.3+0x48576d)
    #3 OBJ_sn2nid crypto/objects/obj_dat.c:615 (libcrypto.so.3+0x4cf5a1)
    #4 evp_pkey_name2type crypto/evp/p_lib.c:983 (libcrypto.so.3+0x44a0aa)
    #5 help_get_legacy_alg_type_from_keymgmt crypto/evp/pmeth_lib.c:162 (libcrypto.so.3+0x45f0d9)
    #6 do_name crypto/core_namemap.c:126 (libcrypto.so.3+0x47ec52)
    #7 doall_util_fn crypto/lhash/lhash.c:207 (libcrypto.so.3+0x47a0ba)
    #8 OPENSSL_LH_doall_arg crypto/lhash/lhash.c:222 (libcrypto.so.3+0x47a1a0)
    #9 lh_NAMENUM_ENTRY_doall_DOALL_NAMES_DATA crypto/core_namemap.c:129 (libcrypto.so.3+0x47eb35)
    #10 ossl_namemap_doall_names crypto/core_namemap.c:141 (libcrypto.so.3+0x47ea7e)
    #11 evp_names_do_all crypto/evp/evp_fetch.c:574 (libcrypto.so.3+0x41ba4b)
    #12 EVP_KEYMGMT_names_do_all crypto/evp/keymgmt_meth.c:277 (libcrypto.so.3+0x435c82)
    #13 get_legacy_alg_type_from_keymgmt crypto/evp/pmeth_lib.c:169 (libcrypto.so.3+0x45f037)
    #14 int_ctx_new crypto/evp/pmeth_lib.c:277 (libcrypto.so.3+0x45310d)
    #15 EVP_PKEY_CTX_new crypto/evp/pmeth_lib.c:440 (libcrypto.so.3+0x4546dd)
    [...]

SUMMARY: ThreadSanitizer: lock-order-inversion (potential deadlock) crypto/threads_pthread.c:89 in CRYPTO_THREAD_write_lock
```

And

```
WARNING: ThreadSanitizer: lock-order-inversion (potential deadlock) (pid=4440)
  Cycle in lock order graph: M786 (0x7b1000002280) => M829 (0x7b1000002ac0) => M1331 (0x7b1000004740) => M786

  Mutex M829 acquired here while holding mutex M786 in main thread:
    [...]
    #1 CRYPTO_THREAD_write_lock crypto/threads_pthread.c:89 (libcrypto.so.3+0x4a7668)
    #2 ossl_provider_find crypto/provider_core.c:230 (libcrypto.so.3+0x49f351)
    #3 provider_conf_load crypto/provider_conf.c:118 (libcrypto.so.3+0x49e9e5)
    #4 provider_conf_init crypto/provider_conf.c:168 (libcrypto.so.3+0x49e5f6)
    #5 module_init crypto/conf/conf_mod.c:438 (libcrypto.so.3+0x2e0cf7)
    #6 module_run crypto/conf/conf_mod.c:294 (libcrypto.so.3+0x2deb82)
    #7 CONF_modules_load crypto/conf/conf_mod.c:193 (libcrypto.so.3+0x2de8ae)
    #8 CONF_modules_load_file_ex crypto/conf/conf_mod.c:236 (libcrypto.so.3+0x2ded4b)
    #9 CONF_modules_load_file crypto/conf/conf_mod.c:258 (libcrypto.so.3+0x2df02b)
    #10 openssl_config_int crypto/conf/conf_sap.c:63 (libcrypto.so.3+0x2e13b9)
    #11 ossl_init_config crypto/init.c:237 (libcrypto.so.3+0x486895)
    #12 ossl_init_config_ossl_ crypto/init.c:235 (libcrypto.so.3+0x485ff2)
    [...]
    #14 CRYPTO_THREAD_run_once crypto/threads_pthread.c:129 (libcrypto.so.3+0x4a7825)
    #15 OPENSSL_init_crypto crypto/init.c:544 (libcrypto.so.3+0x4857a5)
    #16 OPENSSL_init_ssl ssl/ssl_init.c:131 (libssl.so.3+0x7b184)
    [...]

    Hint: use TSAN_OPTIONS=second_deadlock_stack=1 to get more informative warning message

  Mutex M1331 acquired here while holding mutex M829 in main thread:
    [...]
    #1 CRYPTO_THREAD_write_lock crypto/threads_pthread.c:89 (libcrypto.so.3+0x4a7668)
    #2 ossl_namemap_add_name_n crypto/core_namemap.c:254 (libcrypto.so.3+0x47f2dc)
    #3 ossl_namemap_add_name crypto/core_namemap.c:265 (libcrypto.so.3+0x47f63d)
    #4 construct_loader crypto/store/store_meth.c:241 (libcrypto.so.3+0x5bcbfb)
    #5 ossl_method_construct_this crypto/core_fetch.c:70 (libcrypto.so.3+0x47e36a)
    #6 algorithm_do_this crypto/core_algorithm.c:65 (libcrypto.so.3+0x47dd9b)
    #7 provider_forall_loaded crypto/provider_core.c:675 (libcrypto.so.3+0x4a2051)
    #8 ossl_provider_forall_loaded crypto/provider_core.c:753 (libcrypto.so.3+0x4a1cdb)
    #9 ossl_algorithm_do_all crypto/core_algorithm.c:109 (libcrypto.so.3+0x47da8a)
    #10 ossl_method_construct crypto/core_fetch.c:124 (libcrypto.so.3+0x47e06a)
    #11 inner_loader_fetch crypto/store/store_meth.c:295 (libcrypto.so.3+0x5bbe52)
    #12 OSSL_STORE_LOADER_fetch crypto/store/store_meth.c:317 (libcrypto.so.3+0x5bbc47)
    #13 OSSL_STORE_attach crypto/store/store_lib.c:940 (libcrypto.so.3+0x5bb3ac)
    #14 pem_read_bio_key crypto/pem/pem_pkey.c:57 (libcrypto.so.3+0x4eb82d)
    #15 PEM_read_bio_PrivateKey_ex crypto/pem/pem_pkey.c:144 (libcrypto.so.3+0x4ebd6a)
    #16 PEM_read_bio_PrivateKey crypto/pem/pem_pkey.c:151 (libcrypto.so.3+0x4ebe0e)
    [...]

  Mutex M786 acquired here while holding mutex M1331 in main thread:
    [...]
    #1 CRYPTO_THREAD_write_lock crypto/threads_pthread.c:89 (libcrypto.so.3+0x4a7668)
    #2 OPENSSL_init_crypto crypto/init.c:542 (libcrypto.so.3+0x48576d)
    #3 OBJ_sn2nid crypto/objects/obj_dat.c:615 (libcrypto.so.3+0x4cf5a1)
    #4 evp_pkey_name2type crypto/evp/p_lib.c:983 (libcrypto.so.3+0x44a0aa)
    #5 help_get_legacy_alg_type_from_keymgmt crypto/evp/pmeth_lib.c:162 (libcrypto.so.3+0x45f0d9)
    #6 do_name crypto/core_namemap.c:126 (libcrypto.so.3+0x47ec52)
    #7 doall_util_fn crypto/lhash/lhash.c:207 (libcrypto.so.3+0x47a0ba)
    #8 OPENSSL_LH_doall_arg crypto/lhash/lhash.c:222 (libcrypto.so.3+0x47a1a0)
    #9 lh_NAMENUM_ENTRY_doall_DOALL_NAMES_DATA crypto/core_namemap.c:129 (libcrypto.so.3+0x47eb35)
    #10 ossl_namemap_doall_names crypto/core_namemap.c:141 (libcrypto.so.3+0x47ea7e)
    #11 evp_names_do_all crypto/evp/evp_fetch.c:574 (libcrypto.so.3+0x41ba4b)
    #12 EVP_KEYMGMT_names_do_all crypto/evp/keymgmt_meth.c:277 (libcrypto.so.3+0x435c82)
    #13 get_legacy_alg_type_from_keymgmt crypto/evp/pmeth_lib.c:169 (libcrypto.so.3+0x45f037)
    #14 int_ctx_new crypto/evp/pmeth_lib.c:277 (libcrypto.so.3+0x45310d)
    #15 EVP_PKEY_CTX_new crypto/evp/pmeth_lib.c:440 (libcrypto.so.3+0x4546dd)
    [...]

SUMMARY: ThreadSanitizer: lock-order-inversion (potential deadlock) crypto/threads_pthread.c:89 in CRYPTO_THREAD_write_lock
```

**Comment by t8m** [comment]
Please retest with the latest master checkout. It could be that this was fixed by #13660

**Comment by jwalch** [comment]
Thanks for the pointer. I'd been following loosely and I did notice that it seems like a bunch of race conditions got fixed via that PR.

Just a heads up: I may wait a while to retest it. It's a bit more convenient to stick with actual releases in the environment in which I can reproduce this, so I would likely wait until alpha11 is formally available (which I expect is not too far off). I do understand that this was tagged with the beta1 milestone, and I certainly would not want resolution of this issue to be something holding that up.
